### PR TITLE
[Feature](bangc-ops):add new op masked_col2im_forward

### PR DESCRIPTION
**Pull request by defei-coder**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add new op masked_col2im_forward

## 2. Modification

Modify files
- bangc-ops/kernel_depends.toml
- bangc-ops/mlu_op.h

Add new files
- bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward.cpp
- bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward.h
- bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward_union1.mlu
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/masked_col2im_forward.cpp
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/masked_col2im_forward.h
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/test_case/case_0.prototxt
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/test_case/case_1.prototxt
- docs/bangc-docs/design_docs/masked_col2im_forward/masked_col2im_forward.md
- docs/bangc-docs/design_docs/masked_col2im_forward/maskedcol2im.png

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff3: diff3 == 0

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |         U1       |
|        3       |Layouts                               |         NCHW    |
|        4       |Whether multi-dimensions are supported|        No                              |
|        5       |Whether element zero is supported     |     mask_h_idx，mask_w_idx和col的第二维支持0元素，返回MLUOP_STATUS_SUCCESS；col的第一维和im不支持0元素，MLUOP_STATUS_BAD_PARAM           |
|        6       |Data type(half/float)                 |           half / float      |
|        7       |Whether there is size limit           |          No                            |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [ ] Multi-dimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| im（输出）的layout必须是NCHW |     Normal error    |     [Error]:[mluOpMaskedCol2imForward] Check failed: im_desc->layout == MLUOP_LAYOUT_NCHW.     |
| col(第一个输入)的维度必须是2          |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: col_desc->dim == 2.     |
| mask_h_idx和mask_w_idx 的维度必须是1          |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: mask_w_idx_desc->dim == 1.    |
| im和col的第一维必须相等          |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: col_desc->dims[0] == im_desc->dims[1].   |
| im和col的第一维必须相等          |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: col_desc->dims[0] == im_desc->dims[1].   |
| col的数据类型必须是half或float         |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: col_desc->dtype == MLUOP_DTYPE_FLOAT || col_desc->dtype == MLUOP_DTYPE_HALF.   |
| col和im的数据类型必须相同         |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: col_desc->dtype == im_desc->dtype.  |
| col的第二维和mask_h_idx、mask_w_idx的第一维必须相同         |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: col_desc->dims[1] == mask_h_idx_desc->dims[0]. |
| mask_h_idx和mask_w_idx的数据类型必须是int32_t         |     Normal error    |    [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: mask_w_idx_desc->dtype == MLUOP_DTYPE_INT32. |
| im的第一维（dims[0]）必须是1         |     Normal error    |   [MLUOP] [Error]:[mluOpMaskedCol2imForward] Check failed: im_desc->dims[0] == 1.  |
| 不支持large tensor  |     Normal error    |  [MLUOP] [Error]:[[mluOpMaskedCol2imForward]]: overflow max supported tensor num. 2147483647, now tensor's total num is 2147483650.  |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float               |      [ SUMMARY  ] Total 209 cases of 1 op(s). ALL PASSED.    |          |
|Mult-tensor test     |Supports 4 dims                 |   [ SUMMARY  ] Total 209 cases of 1 op(s). ALL PASSED.       |          |
|Layout test          |Supports NCHW               |     [ SUMMARY  ] Total 209 cases of 1 op(s). ALL PASSED.     |          |
|Zero element test    |Whether to support this test      |    col的第一维和im不支持0元素:  [MLUOP] [Error]:[mluOpMaskedCol2imForward] Zero element tensor failure.     |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |   [ SUMMARY  ] Total 211 cases of 1 op(s). ALL PASSED.        |          |
|Nan/INF test         |Whether to support this test      |  yes        |          |
|Memory leak check    |      |  [ SUMMARY  ] Total 211 cases of 1 op(s). ALL PASSED.    |          |
|Code coverage check  |  |  masked_col2im_forward_union1.mlu [94.4%] 94.4 %67 / 71 100.0 %5 / 5   |   代码量少，未执行代码为switch语句的default代码块    |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|masked_col2im_forward| 985 | 304 | 0.000676819 |   -1 |  614400 (Bytes)  |  float  | col:[256,200], im:[1,256,20,20] |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|masked_col2im_forward| 80 | 144.65 | 0.000925926 | -1 | 614400 (Bytes) |  float  | col:[256,200], im:[1,256,20,20] |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/bangc-docs/design_docs/masked_col2im_forward/masked_col2im_forward.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
+# masked_col2im_forward 算子开发设计方案
+
+
+- #### 文档基本信息
+
+| 算子名称    | masked_col2im_forward                                                      |
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/masked_col2im_forward/masked_col2im_forward.md:6]: 这个表乱码了，& 补充9_index.rst

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/masked_col2im_forward.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  int64_t getTheoryIoSize() override;
+
+ private:
+  void init();
+  void printDataInfo();
+  int batchs = 0;
+  int height = 0;
+  int width = 0;
+  int channels = 0;
+  int mask_cnt = 0;
```
Comment (WangYuan-cn) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/masked_col2im_forward.h:48]: 风格统一改成  name_;

File: bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward_union1.mlu [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ *************************************************************************/
+#include "masked_col2im_forward.h"
+
+#include <algorithm>
+
+#include "kernels/debug.h"
+#include "kernels/kernel.h"
+#include "kernels/utils/common.h"
+#include "core/logging.h"
+__nram__ char data_nram[MAX_NRAM_SIZE];
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward_union1.mlu:31]: 头文件、函数和宏定义隔开吧

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -11893,6 +11893,153 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * The descriptor of the tensor \b mask_h_idx. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] mask_w_idx_desc
+ * Descriptor of input data \b mask_w_idx. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] im_desc
+ * Descriptor of input data \b im. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[out] workspace_size
+ * The size of the extra workspace in bytes.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_NOT_SUPPORTED
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:11920]: 代码里没看到 MLUOP_STATUS_NOT_SUPPORTED

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -11893,6 +11893,153 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * @par Return
+ * - MLUOP_STATUS_SUCCESS, MLUOP_STATUS_BAD_PARAM, MLUOP_STATUS_NOT_SUPPORTED
+ *
+ * @par Data Type
+ * - This function supports the following data types for \b col tensor, \b mask_h_idx tensor,
+ *   \b mask_w_idx tensor and \b im tensor.
+ *   Data types of col tensor and im tensor should be the same.
+ *   - col tensor: half, float.
+ *   - mask_h_idx tensor: int32_t.
+ *   - mask_w_idx tensor: int32_t.
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:11992]: ops 风格不加 '.'
```suggestion
 *   - mask_w_idx tensor: int32_t
```

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/masked_col2im_forward.h (hunk 2 above)
Comment (defei-coder) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/masked_col2im_forward/masked_col2im_forward.h:48]: 已修改

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:11920]: 已修改

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:11992]: 已修改

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -11893,6 +11893,153 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * - When the element number of \b im tensor equals to zero, this function will return MLUOP_STATUS_BAD_PARAM.
+ * - When size of the highest dimension of \b col tensor equals to zero, this function will return
+ *   MLUOP_STATUS_BAD_PARAM.
+ *
+ * @par API Dependency
+ * - Before calling this function you need to call ::mluOpGetMaskedCol2imForwardWorkspaceSize
+ *   to get the extra space size needed in ::mluOpMaskedCol2imForward operation.
+ *
+ * @par Note
+ * - The data of \b mask_h_idx should be in the range of [0, h - 1], h is the height of \b im tensor.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:12019]: 【全文排查一下】
```suggestion
 * - The data of \b mask_h_idx must be in the range of [0, h - 1], and h is the height of tnesor \b im .
```

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -11893,6 +11893,153 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * - The data of \b mask_w_idx should be in the range of [0, w - 1], w is the width of \b im tensor.
+ * - The coordinates by combining \b mask_h_idx tensor and \b mask_w_idx tensor can not repeate.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:12021]: repeate？

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -11893,6 +11893,153 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *
+ * @par Example
+ * - None.
+ *
+ * @par Reference
+ * - https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/csrc/pytorch/cuda/masked_conv2d_cuda.cu
+ */
+
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:12029]: 建议删除多余空行

File: bangc-ops/mlu_op.h (hunk 7 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 8 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h]: 已修改

**Review by AquamarineSu — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]

**Review by PetrelYy — APPROVED** [review]

**Review by PetrelYy — APPROVED** [review]

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 9]
```diff
@@ -295,6 +295,13 @@ Llg10的计算公式为：
 - ``i`` 表示一个多元数组的索引，表示多维张量。
 - :math:`x_i`、:math:`y_i` 表示多元组中 i 索引处的元素。
 
+.. __masked_col2im_forward:
```
Comment (duzekunKTH) [inline docs/bangc-docs/user_guide/9_operators/index.rst:298]: 下划线只有一格

File: bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward_union1.mlu [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    const void *mask_h_idx, const void *mask_w_idx, const int mask_cnt,
+    void *im) {
+  if (__is_mpu()) {
+    return;
+  }
+  MLUMultiKernelMaskedCol2imForward((T *)col, height, width, channels,
+                                    (int32_t *)mask_h_idx,
+                                    (int32_t *)mask_w_idx, mask_cnt, (T *)im);
+}
+void MLUOP_WIN_API KernelMaskedCol2imForward(
```
Comment (duzekunKTH) [inline bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward_union1.mlu:98]: 空行

File: docs/bangc-docs/user_guide/9_operators/index.rst (hunk 9 above)
Comment (defei-coder) [reply docs/bangc-docs/user_guide/9_operators/index.rst]: 已修改

File: bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward_union1.mlu (hunk 10 above)
Comment (defei-coder) [reply bangc-ops/kernels/masked_col2im_forward/masked_col2im_forward_union1.mlu:99]: 已修改

**Review by duzekunKTH — APPROVED** [review]

**Pull request merged**